### PR TITLE
373 Unit Tests

### PR DESCRIPTION
**Pull request by oliverweissl**
unit tests for examples.
 designed by @dangbert

### CODE REVIEW

**Comment by surgura** [comment]
can you add the test requirements to the dev install script

File: tests/pyproject.toml
Comment (surgura) [inline tests/pyproject.toml:1]: I don't think the tests need to be a package. the requirements.txt are okay on their own

File: tests/examples/test_simulate_single_robot.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+
+def test_simulate_single_robot() -> None:
+    """Test simulate_single_robot example can complete."""
+    exp_dir = os.path.join(EXAMPLES_DIR, "simulate_single_robot")
+    assert_command_succeeds(
+        [
+            "python",
+            os.path.join(exp_dir, "main.py"),
+            "-t",
```
Comment (surgura) [inline tests/examples/test_simulate_single_robot.py:13]: I don't think this flag still exists. However, speeding up sim to 1 seconds(i think that is what is intended here) is nice. maybe patch the `make_standard_batch_parameters`?

File: tests/examples/test_robot_bodybrain_ea_database.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+def test_robot_bodybrain_ea_database(mocker: Mock, tmpdir: str) -> None:
+    """
+    Test if robot_bodybrain_ea_database example can complete.
+
+    The database file is written to a temporary directory, which is automatically deleted after the test.
+
+    :param mocker: The mock object.
+    :param tmpdir: Temporary directory for spawning database file.
+    """
+    example_dir = os.path.join(EXAMPLES_DIR, "robot_bodybrain_ea_database")
```
Comment (surgura) [inline tests/examples/test_robot_bodybrain_ea_database.py:18]: example_dir should be named differently

File: tests/examples/test_robot_bodybrain_ea_database.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+    # Clear any previously imported modules from examples directory from cache
+    clear_example_modules_from_cache()
+
+    # Add examples directory to path, so we can import them without the examples being packages.
```
Comment (surgura) [inline tests/examples/test_robot_bodybrain_ea_database.py:23]: and then here 'example directory' should also reflect that change

File: tests/examples/test_robot_bodybrain_ea_database.py (hunk 2 above)
Comment (surgura) [reply tests/examples/test_robot_bodybrain_ea_database.py:18]: exp_dir? like with the other tests?

File: tests/examples/test_robot_bodybrain_ea_database.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    mocker.patch("config.NUM_REPETITIONS", 1)
+    mocker.patch("config.POPULATION_SIZE", 4)
+    mocker.patch("config.OFFSPRING_SIZE", 2)
+    mocker.patch("config.NUM_GENERATIONS", 2)
+
+    # Import the example main and run it.
+    try:
+        import main  # type: ignore
+
+        """This type ignore is required since mypy cant resolve this import."""
```
Comment (surgura) [inline tests/examples/test_robot_bodybrain_ea_database.py:38]:
```suggestion
        # This type ignore is required since mypy cant resolve this import.
        import main  # type: ignore
```

File: tests/examples/test_robot_brain_cmaes.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+from ._clear_example_modules_from_cache import clear_example_modules_from_cache
+
+
+def test_robot_brain_cmaes(mocker: Mock) -> None:
+    """
+    Test robot_brain_cmaes example can complete.
+
+    :param mocker: The mock object.
+    """
+    example_dir = os.path.join(EXAMPLES_DIR, "robot_brain_cmaes")
```
Comment (surgura) [inline tests/examples/test_robot_brain_cmaes.py:15]: see comments in other test

File: tests/examples/test_robot_brain_cmaes.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+    # Clear any previously imported modules from examples directory from cache
+    clear_example_modules_from_cache()
+
+    # Add examples directory to path, so we can import them without the examples being packages.
```
Comment (surgura) [inline tests/examples/test_robot_brain_cmaes.py:20]: see comments in other test

File: tests/examples/test_robot_brain_cmaes.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    sys.path.insert(0, example_dir)  # Add the example directory to sys.path
+
+    # Override default config to reduce number of generations.
+    mocker.patch("config.NUM_GENERATIONS", 2)
+
+    # Import the example main and run it.
+    try:
+        import main  # type: ignore
+
+        """This type ignore is required since mypy cant resolve this import."""
```
Comment (surgura) [inline tests/examples/test_robot_brain_cmaes.py:30]:
```suggestion
        # This type ignore is required since mypy cant resolve this import.
        import main  # type: ignore
```

**Review by surgura — APPROVED** [review]
very good

File: tests/pyproject.toml
Comment (surgura) [reply tests/pyproject.toml:1]: the init is fine if that is needed for the test to run. i'm talking about the pyproject.toml specifically. no one is supposed to actually install the tests as a package i think

Comment (oliverweissl) [reply tests/pyproject.toml:1]: this was a simple fix to mypy complaining in the ci.
Since i add tests dir in the packages.txt to also make the checkall find it. i can change this aswell.

File: tests/examples/test_simulate_single_robot.py (hunk 1 above)
Comment (oliverweissl) [reply tests/examples/test_simulate_single_robot.py:13]: im not sure how to approach this. wouldnt i have to pass something to the script?

File: .github/workflows/main.yml [hunk 8]
```diff
@@ -87,7 +87,7 @@ jobs:
         if: startsWith(matrix.package, 'examples/')
       - name: Install package
         run: pip install ./$(echo "${{ matrix.package }}[dev]" | sed 's#/revolve2##')
-        if: "!startsWith(matrix.package, 'examples/')"
+        if: "!startsWith(matrix.package, 'examples/') && !startsWith(matrix.package, 'tests/')"
```
Comment (oliverweissl) [inline .github/workflows/main.yml:90]: this one doesnt work as a expected. Is this python?

File: tests/examples/_patched_batch_parameters.py [hunk 9]
```diff
@@ -0,0 +1,17 @@
+"""Standard simulation functions and parameters patched for unit test."""
+# This is required, since the CI does not resolve those stubs correctly.
+from revolve2.simulation.simulator import BatchParameters  # mypy: allow-untyped-defs
```
Comment (oliverweissl) [inline tests/examples/_patched_batch_parameters.py:3]: This doesnt work with the CI -> type: ignore throws error locally

Comment (oliverweissl) [reply tests/examples/_patched_batch_parameters.py]: adjusted mypy.ini which works, but probably not optimal

File: codetools/mypy/mypy.ini [hunk 10]
```diff
@@ -41,3 +38,6 @@ ignore_missing_imports = True
 
 [mypy-setuptools.*]
 ignore_missing_imports = True
+
+[mypy-revolve2.simulation.*]
```
Comment (surgura) [inline codetools/mypy/mypy.ini:42]: what's up with this? why is this here

File: .github/workflows/main.yml [hunk 11]
```diff
@@ -87,7 +87,7 @@ jobs:
         if: startsWith(matrix.package, 'examples/')
       - name: Install package
         run: pip install ./$(echo "${{ matrix.package }}[dev]" | sed 's#/revolve2##')
-        if: "!startsWith(matrix.package, 'examples/')"
+        if: "(!startsWith(matrix.package, 'examples/') && !startsWith(matrix.package, 'tests'))"
```
Comment (surgura) [inline .github/workflows/main.yml:90]: are the outer brackets needed?

**Review by surgura — CHANGES_REQUESTED** [review]

File: codetools/mypy/mypy.ini (hunk 10 above)
Comment (oliverweissl) [reply codetools/mypy/mypy.ini:42]: The CI fails for some reason at mypy tests, wince it cant resolve the subs for this package

File: .github/workflows/main.yml (hunk 11 above)
Comment (oliverweissl) [reply .github/workflows/main.yml:90]: yes without it didnt work for some reason

Comment (surgura) [reply .github/workflows/main.yml:90]: 🤷

**Review by surgura — APPROVED** [review]
